### PR TITLE
A generic version of ManipulationAgent

### PR DESCRIPTION
**Pull request by zephirefaith**
It has no pre-conceived notions about which object to detect and which task to do until populated by the language planner. 

## Changelog

- New environment which takes in goals as they come (in slap_manipulation but can move to src if it looks useful)
- New agent with a new `GenericTaskState`
- Adds a method in `DeticPerception` to reset vocab on-the-fly
- Also improves Xbox controller so motion is smoother
  - One downside: For precise control on smallest amount of motion we sacrifice overall max speed of translation

## Types of changes

- [x] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

# Checklist:

- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [x] I ran `tests/hw_manual_test.py` on hardware and observed that robot behavior looks normal.

### CODE REVIEW

**Comment by zephirefaith** [comment]
@exhaustin @cpaxton My tests are failing due to linting. I can't reproduce these errors on my desktop even after installing the exact versions as written in `pre-commit` config.

**Comment by exhaustin** [comment]
> @exhaustin @cpaxton My tests are failing due to linting. I can't reproduce these errors on my desktop even after installing the exact versions as written in `pre-commit` config.

Did you rerun pre-commit with `pre-commit run --all-files` ?
I think precommit uses it's own environment so manually installing the correct versions of isort won't help.

File: projects/data_collection/collect_h5.py [hunk 1]
```diff
@@ -79,7 +79,7 @@ def record_keyframe(self):
 @click.option("--dir-path", default="./H5s/", help="Path of root data directory")
 def main(task_name, dir_path):
     em = EpisodeManager(task_name, dir_path)
-    rate = rospy.Rate(10)
+    rate = rospy.Rate(20)
```
Comment (cpaxton) [inline projects/data_collection/collect_h5.py:82]: why?

File: projects/slap_manipulation/readme.md [hunk 2]
```diff
@@ -18,8 +18,14 @@
...
-  - [ ] Edit to respect new simple API
+  - [x] Load pre-trained weights and run initial tests (need new weights for these; queued for after trainig)
+  - [x] Edit to respect new simple API
 - [ ] --> Collect new data from stretch
 - [ ] Edit ported dataloaders to visualize and train using collected H5s
 
+## Natural Language Commands to Steps
+```
+cd scripts
+python get_cmd_to_steps_data.py
```
Comment (cpaxton) [inline projects/slap_manipulation/readme.md:29]: at some point we should document how to run the LLM...

File: projects/slap_manipulation/scripts/create_new_h5.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import glob
+import os
+
+import click
+import h5py
+
+
+@click.command()
+@click.option("--new_file_name", type=click.Path())
+def main(new_file_name):
```
Comment (cpaxton) [inline projects/slap_manipulation/scripts/create_new_h5.py:10]: can you add a brief note about what this script is for?

File: projects/slap_manipulation/scripts/create_new_h5.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    h5_files = glob.glob(os.path.join(data_dir, "*.h5"))
+
+    # Create a new h5 file
+    new_h5_file = h5py.File(new_file_name, "w")
+
+    # For each h5 file in the list
+    for i, h5_file in enumerate(h5_files):
+        print(f"{h5_file=}")
+        breakpoint()
+        new_h5_file[f"/0_{i}"] = h5py.ExternalLink(h5_file, "/0")
```
Comment (cpaxton) [inline projects/slap_manipulation/scripts/create_new_h5.py:22]: interesting... also did you mean to commit a breakpt here?

File: projects/slap_manipulation/scripts/eval_lang_agent.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
+import click
+import rospy
+from slap_manipulation.agents.language_ovmm_agent import LangAgent
+from slap_manipulation.env.language_planner_env import LanguagePlannerEnv
+
+from home_robot_hw.env.stretch_pick_and_place_env import (
+    REAL_WORLD_CATEGORIES,
```
Comment (cpaxton) [inline projects/slap_manipulation/scripts/eval_lang_agent.py:7]: can you create a separate copy of this for your use case?

File: projects/slap_manipulation/src/slap_manipulation/agents/language_ovmm_agent.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+import pandas as pd
+
+from home_robot.agent.ovmm_agent.pick_and_place_agent import (
+    PickAndPlaceAgent,
+    SimpleTaskState,
+)
+from home_robot.core.interfaces import Action, DiscreteNavigationAction, Observations
+
+
+class GeneralTaskState(Enum):
```
Comment (cpaxton) [inline projects/slap_manipulation/src/slap_manipulation/agents/language_ovmm_agent.py:15]: do you want to move this to home_robot.core.interfaces or something? feel like it's pretty general

File: projects/slap_manipulation/src/slap_manipulation/agents/language_ovmm_agent.py [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+    NOT_STARTED = 0
+    PREPPING = 1
+    DOING_TASK = 2
+    IDLE = 3
+    STOP = 4
+
+
+def get_task_plans_from_gt(
```
Comment (cpaxton) [inline projects/slap_manipulation/src/slap_manipulation/agents/language_ovmm_agent.py:23]: from gt = from ground truth?

Comment (cpaxton) [reply projects/slap_manipulation/src/slap_manipulation/agents/language_ovmm_agent.py:23]: technical term would be oracle

File: projects/slap_manipulation/src/slap_manipulation/agents/language_ovmm_agent.py [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+    # steps_df = pd.DataFrame.from_records(steps_list)
+    code = get_codelist(steps_list)
+    return code
+
+
+def get_codelist(steps_list):
+    codelist = []
+    for step in steps_list:
+        codelist += [
+            f"self.{step['verb']}('{step['noun']}', motion_profile={step['adverb']}, obs=obs)"
```
Comment (cpaxton) [inline projects/slap_manipulation/src/slap_manipulation/agents/language_ovmm_agent.py:46]: these are the specific commands to execute right?

File: projects/slap_manipulation/src/slap_manipulation/dataloaders/robot_loader.py [hunk 9]
```diff
@@ -262,6 +263,17 @@ def get_gripper_pose(self, trial, idx):
         ee_pose = ee_pose @ self._robot_ee_to_grasp_offset
         return ee_pose
 
+    def get_gripper_axis(self, rot_mat):
+        return rot_mat[:3, 2]
+
+    def find_closest_point_to_line(self, xyz, gripper_position, action_axis):
```
Comment (cpaxton) [inline projects/slap_manipulation/src/slap_manipulation/dataloaders/robot_loader.py:269]: can this go in a better place? utils?

File: projects/slap_manipulation/src/slap_manipulation/dataloaders/robot_loader.py [hunk 10]
```diff
@@ -262,6 +263,17 @@ def get_gripper_pose(self, trial, idx):
         ee_pose = ee_pose @ self._robot_ee_to_grasp_offset
         return ee_pose
 
+    def get_gripper_axis(self, rot_mat):
```
Comment (cpaxton) [inline projects/slap_manipulation/src/slap_manipulation/dataloaders/robot_loader.py:266]: maybe we need robot utils for this

File: projects/slap_manipulation/src/slap_manipulation/dataloaders/robot_loader.py [hunk 11]
```diff
@@ -466,6 +478,22 @@ def get_datum(self, trial, keypoint_idx, verbose=False):
...
             cmd = cmd_opts[np.random.randint(len(cmd_opts))]
+        if self.per_action_cmd:
+            # hard-coded for pick_up_bottle right now
+            cmd = [
+                "approach-pose-action bottle",
+                "grasp-action bottle",
+                "lift-action bottle",
+            ]
+        # TODO: remove this and read from a yaml file instead
+        if "knob" in cmd:
```
Comment (cpaxton) [inline projects/slap_manipulation/src/slap_manipulation/dataloaders/robot_loader.py:489]: this is a bit confusing, do tyou mean handle?

File: src/home_robot/home_robot/agent/ovmm_agent/ppo_agent.py [hunk 12]
```diff
@@ -294,6 +294,7 @@ def convert_to_habitat_obs_space(
             }
         )
 
+    # FIXME: the return values do not match the signature
```
Comment (cpaxton) [inline src/home_robot/home_robot/agent/ovmm_agent/ppo_agent.py:297]: @yvsriram @JHurricane96 - does this seem right to you?

File: src/home_robot/home_robot/perception/detection/detic/detic_perception.py [hunk 13]
```diff
@@ -181,6 +181,25 @@ def __init__(
         self.predictor = DefaultPredictor(cfg)
         reset_cls_test(self.predictor.model, classifier, num_classes)
 
+    def reset_vocab(self, new_vocab: List[str], vocab_type="custom"):
```
Comment (cpaxton) [inline src/home_robot/home_robot/perception/detection/detic/detic_perception.py:184]: add docstring, but thanks for adding this, super helpful

**Review by cpaxton — CHANGES_REQUESTED** [review]
Mostly minor comments, some cases where I think code formatting is causing the linter to fail - also please don't commit changes to hab-lab unless they are intenetional

otherwise looks good to me! lots of super helpful stuff, we should merge this

File: src/home_robot/home_robot/agent/ovmm_agent/ppo_agent.py (hunk 12 above)
Comment (yvsriram) [reply src/home_robot/home_robot/agent/ovmm_agent/ppo_agent.py:297]: Yes, this is right. Will fix this one in our integration PR today

File: projects/slap_manipulation/src/slap_manipulation/agents/language_ovmm_agent.py (hunk 8 above)
Comment (zephirefaith) [reply projects/slap_manipulation/src/slap_manipulation/agents/language_ovmm_agent.py:46]: @vidhiJain

File: projects/slap_manipulation/readme.md (hunk 2 above)
Comment (zephirefaith) [reply projects/slap_manipulation/readme.md:29]: @vidhiJain

File: projects/data_collection/collect_h5.py (hunk 1 above)
Comment (zephirefaith) [reply projects/data_collection/collect_h5.py:82]: Trial-and-error for smoother robot execution until I found the fix with Sam.

**Comment by zephirefaith** [comment]
@exhaustin Yes I am using the pre-commit command to format all files. It thinks everything is just fine.

File: src/home_robot_hw/home_robot_hw/nodes/contact_graspnet_server.py [hunk 14]
```diff
@@ -25,12 +25,13 @@
...
 import config_utils  # noqa: E402
 from contact_grasp_estimator import GraspEstimator  # noqa: E402
+from visualization_utils import show_image, visualize_grasps  # noqa: E402
+
 from data import (  # noqa: E402
     depth2pc,
     load_available_input_data,
     regularize_pc_point_count,
 )
-from visualization_utils import show_image, visualize_grasps  # noqa: E402
```
[the remark this answers is not in the document]
Comment (yvsriram) [reply src/home_robot_hw/home_robot_hw/nodes/contact_graspnet_server.py:33]: Yes, this is the reason. I also observed before that: pre-commit run --all-files makes these changes but the tests fail with this. Tests will pass if these changes are reverted.

File: src/home_robot/home_robot/perception/detection/lseg/modules/lsegmentation_module.py [hunk 15]
```diff
@@ -11,14 +11,15 @@
...
 import torch.cuda.amp as amp
-from data import get_available_datasets, get_dataset
 from encoding.nn import SegmentationLosses
 from encoding.utils import (
     SegmentationMetric,
     batch_intersection_union,
     batch_pix_accuracy,
 )
 
+from data import get_available_datasets, get_dataset
```
Comment (yvsriram) [inline src/home_robot/home_robot/perception/detection/lseg/modules/lsegmentation_module.py:21]: This is another change that needs to be reverted for the tests to pass. I think something is off between pre-commit we are using locally and pre-commit CircleCI is using

Comment (yvsriram) [reply src/home_robot/home_robot/perception/detection/lseg/modules/lsegmentation_module.py:21]: Reverting these two files worked for me in the past:
```
src/home_robot/home_robot/perception/detection/lseg/modules/lsegmentation_module.py
src/home_robot_hw/home_robot_hw/nodes/contact_graspnet_server.py
```

Comment (zephirefaith) [reply src/home_robot/home_robot/perception/detection/lseg/modules/lsegmentation_module.py:21]: Yep, I think that's it. I tried `pre-commit clean/gc/uninstall` and then reinstallating the venvs but I see the same linting and formatting behavior which does not match the expected output by the CI 😭

Comment (zephirefaith) [reply src/home_robot/home_robot/perception/detection/lseg/modules/lsegmentation_module.py]: I ended up uninstalling pre-commit to make this work. If I checked out those two files pre-commit would again reformat them to break tests.

File: projects/slap_manipulation/src/slap_manipulation/dataloaders/robot_loader.py (hunk 10 above)
Comment (zephirefaith) [reply projects/slap_manipulation/src/slap_manipulation/dataloaders/robot_loader.py:266]: Straightforward code, made it a function so it is explicit what is happening. Going to let it remain here for now. Will move if I find a good place.

File: projects/slap_manipulation/src/slap_manipulation/dataloaders/robot_loader.py (hunk 11 above)
Comment (zephirefaith) [reply projects/slap_manipulation/src/slap_manipulation/dataloaders/robot_loader.py]: This was a small test; I meant microwave knob, I was testing different interaction points. Removed this, will read from YAML in next iteration.

File: projects/slap_manipulation/src/slap_manipulation/dataloaders/robot_loader.py (hunk 10 above)
Comment (cpaxton) [reply projects/slap_manipulation/src/slap_manipulation/dataloaders/robot_loader.py:267]: Why not put it in Stretch Model?

File: projects/slap_manipulation/src/slap_manipulation/env/language_planner_env.py [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+        self.segmentation.reset_vocab(vocab)
+
+        if len(info["object_list"]) > 1:
+            self.current_goal_id = 2
+            self.current_goal_name = info["object_list"][1]
+        else:
+            self.current_goal_id = 1
+            self.current_goal_name = info["object_list"][0]
+
+    def _switch_to_manip_mode(self, grasp_only=False):
```
Comment (cpaxton) [inline projects/slap_manipulation/src/slap_manipulation/env/language_planner_env.py:69]: Why is this duplicated?

File: projects/slap_manipulation/src/slap_manipulation/env/language_planner_env.py [hunk 17]
```diff
@@ -0,0 +1,250 @@
...
+            test_grasping,
+            dry_run,
+            debug,
+            *args,
+            **kwargs
+        )
+        self.current_goal_id = None
+        self.current_goal_name = None
+
+    def reset(self):
```
Comment (cpaxton) [inline projects/slap_manipulation/src/slap_manipulation/env/language_planner_env.py:46]: is this duplicated from pick and place env?

File: projects/slap_manipulation/src/slap_manipulation/dataloaders/robot_loader.py (hunk 10 above)
Comment (zephirefaith) [reply projects/slap_manipulation/src/slap_manipulation/dataloaders/robot_loader.py:267]: You mean `HelloStretchKinematics`?

Comment (zephirefaith) [reply projects/slap_manipulation/src/slap_manipulation/dataloaders/robot_loader.py:267]: Or maybe `manip.py` is a better place?

Comment (zephirefaith) [reply projects/slap_manipulation/src/slap_manipulation/dataloaders/robot_loader.py:267]: I have more of such methods which I'd like to move to a robot-specific file actually (IMO get gripper pose for example shd use robot-specific calls). I'll handle this in next PR?

File: projects/slap_manipulation/src/slap_manipulation/env/language_planner_env.py (hunk 16 above)
Comment (zephirefaith) [reply projects/slap_manipulation/src/slap_manipulation/env/language_planner_env.py:69]: A lot of differences in behavior. I didn't want to cause regressions in OVMM code with my changes. I am doing away with any notion of knowing `start_recep`, `object` and `goal_recep` at the begining.

File: projects/slap_manipulation/src/slap_manipulation/env/language_planner_env.py [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+                    "[Env] Action not implemented in pick-and-place environment:",
+                    action,
+                )
+                continuous_action = None
+        elif action.is_navigation():
+            continuous_action = action.get()
+        elif action.is_manipulation():
+            joints_action, continuous_action = action.get()
+
+        # Move, if we are not doing anything with the arm
```
[the remark this answers is not in the document]
Comment (zephirefaith) [reply projects/slap_manipulation/src/slap_manipulation/env/language_planner_env.py:163]: Yes, small changes to allow info to carry more than just visualization info.

File: projects/slap_manipulation/src/slap_manipulation/env/language_planner_env.py [hunk 19]
```diff
@@ -0,0 +1,250 @@
...
+            if not self.dry_run:
+                print("GOTO", continuous_action)
+                self.robot.nav.navigate_to(
+                    continuous_action, relative=True, blocking=True
+                )
+        self._handle_joints_action(joints_action)
+        self._handle_gripper_action(gripper_action)
+        return False
+
+    def get_observation(self) -> Observations:
```
[the remark this answers is not in the document]
Comment (zephirefaith) [reply projects/slap_manipulation/src/slap_manipulation/env/language_planner_env.py:178]: As I added above, I have a number of small changes in this env. Would you rather I integrate these in OVMM Env and test? I think that'd take much longer and I do not want to cause regressions in OVMM code.

[the remark this answers is not in the document]
Comment (zephirefaith) [reply projects/slap_manipulation/src/slap_manipulation/env/language_planner_env.py:178]: As I added above, I have a number of small changes in this env. Would you rather I integrate these in OVMM Env and test? I think that'd take much longer and I do not want to cause regressions in OVMM code.

File: projects/slap_manipulation/src/slap_manipulation/dataloaders/robot_loader.py (hunk 10 above)
Comment (cpaxton) [reply projects/slap_manipulation/src/slap_manipulation/dataloaders/robot_loader.py:267]: i think HelloStretchKinematics is better, nothing here that requires ROS right?

but yeah create issues for it

File: projects/slap_manipulation/src/slap_manipulation/env/language_planner_env.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (cpaxton) [reply projects/slap_manipulation/src/slap_manipulation/env/language_planner_env.py:178]: i  think its ok, i don't love how similar the code is

can you add an issue to de-duplicate this?

**Review by cpaxton — APPROVED** [review]
lgtm once issues are created for de-duplication of agent/env code

**Pull request merged**